### PR TITLE
WIP: Socket.io replaces polling

### PR DESCRIPTION
**Pull request by ameliavoncat**

### CODE REVIEW

**Comment by AbrahamFergie** [comment]
One known issue that we're not sure how to track down: we're sporadically getting the following:

```sh
REQUEST:  { host: 'api.github.com',
  port: 443,
  path: '/repos/GuildCrafts/prrr/pulls/50',
  method: 'get',
  headers:
   { host: 'api.github.com',
     'content-length': '0',
     Authorization: 'token fb3639e0ea2fd70377a085afc0f27a913e213c41',
     'user-agent': 'prrr-learnersguild-org',
     accept: 'application/vnd.github.v3.raw' },
  ca: undefined }
STATUS: 401
HEADERS: {"server":"GitHub.com","date":"Wed, 11 Jan 2017 22:23:50 GMT","content-type":"application/json; charset=utf-8","content-length":"83","connection":"close","status":"401 Unauthorized","x-github-media-type":"github.v3; param=raw","x-ratelimit-limit":"60","x-ratelimit-remaining":"54","x-ratelimit-reset":"1484173741","access-control-expose-headers":"ETag, Link, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval","access-control-allow-origin":"*","content-security-policy":"default-src 'none'","strict-transport-security":"max-age=31536000; includeSubdomains; preload","x-content-type-options":"nosniff","x-frame-options":"deny","x-xss-protection":"1; mode=block","x-github-request-id":"D0421852:54F9:9F61451:5876B076"}
[error] { [Error: {"message":"Bad credentials","documentation_url":"https://developer.github.com/v3"}]
[error]   message: '{"message":"Bad credentials","documentation_url":"https://developer.github.com/v3"}',
[error]   code: 401,
[error]   status: 'Unauthorized',
[error]   headers:
[error]    { server: 'GitHub.com',
[error]      date: 'Wed, 11 Jan 2017 22:23:50 GMT',
[error]      'content-type': 'application/json; charset=utf-8',
[error]      'content-length': '83',
[error]      connection: 'close',
[error]      status: '401 Unauthorized',
[error]      'x-github-media-type': 'github.v3; param=raw',
[error]      'x-ratelimit-limit': '60',
[error]      'x-ratelimit-remaining': '54',
[error]      'x-ratelimit-reset': '1484173741',
[error]      'access-control-expose-headers': 'ETag, Link, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval',
[error]      'access-control-allow-origin': '*',
[error]      'content-security-policy': 'default-src \'none\'',
[error]      'strict-transport-security': 'max-age=31536000; includeSubdomains; preload',
[error]      'x-content-type-options': 'nosniff',
[error]      'x-frame-options': 'deny',
[error]      'x-xss-protection': '1; mode=block',
[error]      'x-github-request-id': 'D0421852:54F9:9F61451:5876B076' } } { owner: 'GuildCrafts', repo: 'prrr', number: 50 } null
```

File: browser/actions/logout.js [hunk 1]
```diff
@@ -1,12 +1,11 @@
...
-import loadSession from './loadSession'
 
 export default function logout() {
   return request('post', '/api/logout')
     .catch(logoutError => {
       state.set({logoutError})
       throw logoutError
     })
-    .then(_ => loadSession())
+    .then(_ => window.location.href = window.location.origin )
```
Comment (deadlyicon) [inline browser/actions/logout.js]: `window.location.href = window.location.origin`  -> `location.reload()`

File: browser/actions/unclaimPrrr.js [hunk 2]
```diff
@@ -1,8 +1,5 @@
-import request from '../request'
-import state from '../state'
-import loadPrrrs from './loadPrrrs'
+import socket from '../socket'
 
 export default function unclaimPrrr(prrrId) {
-  return request('post', `/api/pull-request-review-requests/${prrrId}/unclaim`)
-    .then(loadPrrrs)
+  socket.emit('unclaimPrrr', prrrId)
```
Comment (deadlyicon) [inline browser/actions/unclaimPrrr.js]: I'd prefer all socket message payloads to be an object. `socket.emit('unclaimPrrr', {prrrId})`

File: browser/components/molecules/PendingPrrrs/index.js [hunk 3]
```diff
@@ -19,16 +19,6 @@ export default class PendingPrrrs extends Component {
 
   claimPrrr(prrr){
     claimPrrr(prrr.id)
-      .then(_ => {
```
Comment (deadlyicon) [inline browser/components/molecules/PendingPrrrs/index.js]: we still need to popup the PR. Lets have the server send back a message over the socket (only to us) when the Prrr is claimed. Out client side handler for that event. (maybe call if `prrrClaimedByYou`) should do this

```js
(prrr) => {
  const url = `https://github.com/${prrr.owner}/${prrr.repo}/pull/${prrr.number}`
  const popup = window.open(url, '_blank')
   if (popup) popup.focus()
}
```

File: browser/index.js [hunk 4]
```diff
@@ -2,6 +2,7 @@ import React from 'react'
 import { render } from 'react-dom'
 import './index.sass'
 import Root from './components/Root'
+import socket from './socket'
```
Comment (deadlyicon) [inline browser/index.js]: we dont need to require anymore. most of the actions require it

File: server/socket.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    })
+
+    socket.on('claimPrrr', prrrId => {
+      commands.markPullRequestAsClaimed(prrrId)
+        .then(prrr => updatePrrr('updatePrrr', prrr))
+    })
+
+    socket.on('unclaimPrrr', prrrId => {
+      commands.unclaimPrrr(prrrId)
+      .then(prrr => updatePrrr('updatePrrr', prrr))
```
Comment (deadlyicon) [inline server/socket.js]: indent

File: server/socket.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      commands.completePrrr(prrrId)
+        .then(prrr => updatePrrr('removePrrr', prrr))
+    })
+
+    socket.on('createPrrr', data => {
+      commands.createPrrr(data)
+        .then(prrr => updatePrrr('newPrrr', prrr))
+    })
+
+    const updatePrrr = ( type, prrr ) => {
```
Comment (deadlyicon) [inline server/socket.js]: I wouldn't call this `updatePrrr`  I'd call it `broadcastToAll` or something like that.

**Comment by deadlyicon** [comment]
I'm refactoring this

**Pull request closed without merging**